storage/repository/repository.go

```diff
@@ -245,3 +248,88 @@ func ForURI(u fyne.URI) (Repository, error) {
 
 	return repo, nil
 }
+
+// ParseURI implements the back-end logic for storage.ParseURI, which you
+// should use instead. This is only here because other functions in repository
+// need to call it, and it prevents a circular import.
+//
+// Since: 2.0.0
+func ParseURI(s string) (fyne.URI, error) {
+	// Extract the scheme.
+	scheme := ""
+	for i := 0; i < len(s); i++ {
+		if s[i] == ':' {
+			break
+		}
+		scheme += string(s[i])
+	}
+
+	if scheme == "file" {
+		// Does this really deserve to be special? In principle, the
+		// purpose of this check is to pass it to NewFileURI, which
+		// allows platform path seps in the URI (against the RFC, but
+		// easier for people building URIs naively on Windows). Maybe
+		// we should punt this to whoever generated the URI in the
+		// first place?
+
+		path := s[5:] // everything after file:
+		if len(path) > 2 && path[:2] == "//" {
+			path = path[2:]
+		}
+
+		// this looks weird, but it makes sure that we still pass
+		// url.Parse()
+		s = NewFileURI(path).String()
+	}
+
+	repo, err := ForURI(&uri{scheme: scheme})
+	crepo, ok := repo.(CanonicalRepository)
+	if !ok || err != nil {
```

Comment:
Consider flipping this to reduce overall nesting.

Also we may to be able to return an error rather than falling back to internal implementation, for example if a uri is not supported

Reply:
@andydotxyz implemented this in ab448b2f9ee9c9cc523735abc71f7d8684d2dfef